server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1000,15 +1048,19 @@ public ZoneId readOptionalZoneId() throws IOException {
      * Reads a set of objects
      */
     public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
-        return readCollection(reader, HashSet::new);
+        return readCollection(reader, HashSet::new, Collections.emptySet());
     }
 
     /**
      * Reads a collection of objects
      */
     private <T, C extends Collection<? super T>> C readCollection(Writeable.Reader<T> reader,
-                                                                  IntFunction<C> constructor) throws IOException {
+                                                                  IntFunction<C> constructor,
+                                                                  C empty) throws IOException {
```

Comment:
Instead of passing in an empty already constructed object (even if it is empty, it would still create a ref right?), I think we could pass in Supplier<C>, then use eg Collections::emptySet above?

Reply:
`Collections#emptySet` refers to a fixed existing object for all invocations, it does not create a new object (the Javadocs do not guarantee this, but suggests that implementations can take this optimization if they want; OpenJDK of course does implement this optimization).

Reply:
I think Ryan's object was the creation of a ref not an object here. I don't think that's an issue though, the JVM will just inline `java.util.Collections#EMPTY_SET` after a few calls to that method anyway.

Reply:
Yes, the invocation will be inlined almost surely, and the reference does not matter because `Collections#EMPTY_SET` is rooted anyway (it's a static field in a core JDK class).